src/clj/fluree/db/json_ld/commit_data.cljc

```diff
@@ -395,34 +395,47 @@
   [flakes]
   (filter ref? flakes))
 
+;; TODO - flakes-size takes considerable time for lg txns, see if can be optimized
+(defn calc-flake-size
+  [add rem]
+  (cond-> 0
+          add (+ (flake/size-bytes add))
+          rem (- (flake/size-bytes rem))))
+
 (defn update-novelty
   ([db add]
    (update-novelty db add []))
 
   ([db add rem]
    (try*
-    (let [ref-add     (ref-flakes add)
-          ref-rem     (ref-flakes rem)
-          flake-count (cond-> 0
-                              add (+ (count add))
-                              rem (- (count rem)))
-          flake-size  (cond-> 0
-                              add (+ (flake/size-bytes add))
-                              rem (- (flake/size-bytes rem)))]
-      (-> db
-          (update-in [:novelty :spot] flake/revise add rem)
-          (update-in [:novelty :post] flake/revise add rem)
-          (update-in [:novelty :opst] flake/revise ref-add ref-rem)
-          (update-in [:novelty :tspo] flake/revise add rem)
-          (update-in [:novelty :size] + flake-size)
-          (update-in [:stats :size] + flake-size)
-          (update-in [:stats :flakes] + flake-count)))
-    (catch* e
-            (log/error (str "Update novelty unexpected error while attempting to updated db: "
-                            (pr-str db) " due to exception: " (ex-message e))
-                       {:add-flakes add
-                        :rem-flakes rem})
-            (throw e)))))
+     (let [flake-count (cond-> 0
+                               add (+ (count add))
+                               rem (- (count rem)))
+           ;; launch futures for parallellism on JVM
+           flake-size  #?(:clj  (future (calc-flake-size add rem))
+                          :cljs (calc-flake-size add rem))
+           post        #?(:clj  (future (flake/revise (get-in db [:novelty :post]) add rem))
+                          :cljs (flake/revise (get-in db [:novelty :post]) add rem))
+           opst        #?(:clj  (future (flake/revise (get-in db [:novelty :opst]) (ref-flakes add) (ref-flakes rem)))
+                          :cljs (flake/revise (get-in db [:novelty :opst]) (ref-flakes add) (ref-flakes rem)))]
+       (-> db
+           (update-in [:novelty :spot] flake/revise add rem)
```

Comment:
Does it hurt anything to do spot and tspo in the background as well? While unlikely, there could be a situation where there are lots of properties on a single subject updated at once that might make updating novelty for these indexes slow.

Reply:
There is some overhead in spawning a new thread, but small. I think it is probably in its most optimal state for almost all datasets so I decided to stop here.

It wouldn't be a huge hit to put them all in threads however if desired. The dataset I was testing already had some "s-p" pairs that were ~1,000 flakes however, and still saw this performance difference at this level. They were mostly ref flakes however. I think the circumstance that could push it would be very small "s-p" pairs, but a large number of flakes, where the 'o' consisted of mostly strings.